# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [204]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [205]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [206]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [207]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Здесь и ниже, мы проведём небольшое дополнительное исследование, которое в дальнейшем поможет нам получить дополнительные выводы к нашему исследованию.

Прежде, чем обрабатывать и заменять пропущенные значения в таблице, попробуем их изучить.
Для этого создадим функцию def которая будет отбирать Только те строки, в которых присутствуют пропущенные (NaN) значения в нашей таблице.
Выведем на экран первые 30 строк с пропущенными данными.

In [208]:
def nans(data): # создаем функцию с перебором по "nan" значениям
    return data[data.isnull().any(axis=1)] # возвращаем строки, в которых присутствуют отсутствующие NaN значения
print(nans(data).head(30)) # выводим на экран первые 30 строк с пропущенными данными.

     children  days_employed  dob_years            education  education_id  \
12          0            NaN         65              среднее             1   
26          0            NaN         41              среднее             1   
29          0            NaN         63              среднее             1   
41          0            NaN         50              среднее             1   
55          0            NaN         54              среднее             1   
65          0            NaN         21              среднее             1   
67          0            NaN         52               высшее             0   
72          1            NaN         32               высшее             0   
82          2            NaN         50               высшее             0   
83          0            NaN         52              среднее             1   
90          2            NaN         35               высшее             0   
94          1            NaN         34               высшее    

Дополнительно выведем на экран все уникальные типы доходов кредитополучателей.\
Это необходимо для того чтобы убедиться в том, что в таблице есть отдельная категория 'безработный'.\
В случае, если бы такой категории не было - можно было бы предположить, что отсутствие значений в таблице выше могло бы соответствовать тому, что данные клиенты являлись безработными.\
Однако данная теория не подтвердилась.\
Отдельный статус безработного присутствует.\
А в нашей полученной таблице - тип дохода у кредитополучателей не имеет никаких явных закономерностей.\
(Есть сотрудники, компаньоны, госслужащие и т.д.)


In [209]:
print(data['income_type'].unique())

['сотрудник' 'пенсионер' 'компаньон' 'госслужащий' 'безработный'
 'предприниматель' 'студент' 'в декрете']


Как мы можем убедиться, функция работает.
После того, как мы вывели результаты на экран, мы обнаружили что в первых 30 строках отсутствие значений в столбцах 'days_employed' и 'total_income' полностью совпадает.
Аналогичные наблюдения получились изучив первые 100 строк.

Полученные данные дают возможность сделать предположение, что отсутствие значений в данных двух столбцах взаимосвязано.

Дальнейшие выводы мы распишем в конце данного проекта в разделе выводов.

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [210]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [211]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [212]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [213]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [214]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [215]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [216]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [217]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [218]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [219]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [220]:
data.duplicated().sum()

71

In [221]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [222]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [223]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [224]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [225]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [226]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Для того чтобы определить, есть ли зависимость между количеством детей и возвратом кредита в срок - для начала сгруппируем данные по столбцу, в котором указано количество детей (children) и посчитаем суммарное количество просрочек по платежам, указанных в столбце (debt).

In [227]:
print(data.groupby('children')['debt'].sum())

children
0    1063
1     444
2     194
3      27
4       4
5       0
Name: debt, dtype: int64


Из полученных данных можно сделать предположение, что существует обратно-пропорциональная зависимость между количеством детей в семье и возвратом кредита в срок.

К примеру:

Количество просроченных платежей у кредитополучателей, у которых нет детей - составляет аж целых 1063.

В тоже время семьи, у которых 5 детей - ни разу не пророчили оплаты по кредитам.

*На первый взгляд можно сделать вывод, что чем больше детей в семье - тем большая вероятность, что кредитополучатели оплатят кредит в срок (без просрочек). 
Однако это не окончательные выводы!*

Для того, чтобы определить верны ли наши предположения, необходимо будет также ознакомится с конверсией этих данных - тоесть отношением целевого показателя к общему. (Количеству просрочек у данной рассматриваемой группы к общему количеству человек в данной группе)

Для определения количества человек, находящихся в каждой и рассматриваемых групп воспользуемся функцией .count()

In [228]:
print(data.groupby('children')['debt'].count())

children
0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: debt, dtype: int64


Из полученных данных видим, что наибольшей группой кредитополучателей являются лица, не имеющие детей совсем. (14091 человек).
В то же время количество семей, имеющих 5 детей, и взявших кредит - всего лишь 9.

Теперь не удивительно, почему так много просрочек по кредиту у первой группы(не имеющих детей) и так мало у последних группы (у которых 4 или 5 детей).

Но, как и было указано ранее, для подведения окончательных итогов - необходимо будет рассмотреть конверсию этих данных. 

Учитывая рекомендацию, рассматриваем конверсию данных - используя функцию .mean().  
    Для применения данной функции - группируем данные по столбцу 'children' и рассчитываем среднее значение в столбце 'debt' 

In [230]:
print(data.groupby('children')['debt'].mean().sort_values())

children
5    0.000000
0    0.075438
3    0.081818
1    0.092346
2    0.094542
4    0.097561
Name: debt, dtype: float64


Рассмотрев конверсию данных мы приходим практически к противоположному выводу, который хотелось сделать изначально.

Согласно полученным данным - лишь **7.5%**  кредитополучателей (**не имеющих детей**) совершают просрочки по кредиту, 
в то же время, если рассмотреть категорию семей **с 4-мя детьми**, то из них почти **9,8%** семей совершают просрочки по кредиту.

**Вывод:** 

1. Исследование выявило, что зависимость между количеством детей и возвратом кредита в срок - имеется!

2. Кредитополучатели (**имеющие детей**), с большей вероятностью совершат просрочку по кредиту, чем кредитополучатели (**не имеющие детей вовсе**).

*Исключением из данных будут семьи с 5 детьми.
Данные семьи не совершали просрочек по платежам, но полученные данные нельзя сравнивать в общем контексте, так как данная категория кредитополучателей составляет ничтожную долю от общего числа кредитополучателей (0,04%).*


Конечно «разбег» между "цифрами" в наших данных мал. **Возможно**, что в рамках имеющихся данных сделать вывод о том, что семьи с детьми относительно семей без оных имеют какие-то задержки в выплатах кредитов нельзя. Слишком уж близки доли, чуть измениться выборка и значения могут измениться в другую сторону. То есть можно сделать вывод, что сколь угодно значимой разницы между группами нет ... но даже небольшие разницы могут существенно сказываться на кредитном портфеле банка.

Однако, если считать, что **любая** разница между группами есть повод для «разделения» классов, то я бы сделал такой вывод: кредитополучатели без детей являются более ответственными, чем клиенты, у которых есть дети. Видимо, наличие ребенка накладывает на клиента дополнительные финансовое бремя (обязательства) и клиенты либо менее ответственно относятся к своим кредитным обязательствам, либо не имеют возможности в срок оплачивать обязательства.

Ну и что ещё можно увидеть в наших данных (агрегация «по count» в этом может помочь), то, что по некоторым группам заемщиков у нас недостаточно данных. Говорят, что выборка не сбалансирована, классы в выборке имеют разные размеры.

И вот ещё какой важный момент – да, на больших суммах и маленькие доли процента выльются в значительные объемы. Но вообще, насколько я понимаю логику кредитного бизнеса, дело не в том, чтобы кому-то не давать денег, а в том, чтобы компенсировать риск выдачи кредитным процентом. То есть, потенциальный должник заплатит более высокий процент за использование кредита. Так что в этой задаче фактически рассматривается вопрос - повышать ли процент заемщику с детьми. И статистика, которая показывает, что человек с 4 детьми просрочит выплату примерно так же, как и с двумя, - она, на мой взгляд, не обосновывает ужесточение кредитных условий для семей с детьми. К тому же тут зафиксировано количество детей только на момент выдачи кредита, никто не гарантирует, что их не станет больше в течение срока выплаты. Таким образом из этих данных можно сделать скорее другой вывод: кредитный портфель бездетных (со сроками выданных кредитов по несколько лет) скорее всего со временем ухудшится по показателю возврата денег (какая-то часть бездетных заведут детей, и у них должников из 7,5% станет 9%). И тогда этот риск можно было бы закладывать заранее (но для этого хорошо было бы понимать, сколько бездетных перейдут в категорию «с детьми» в течение 5 лет, например).


#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Для того чтобы определить, есть ли зависимость между семейным положением и возвратом кредита в срок - необходимо будет для начала сгруппировать данные по столбцу, в котором указано семейное положение (family_status) и посчитать суммарное количество просрочек по платежам, указанных в столбце (debt). Для наглядности, полученные данные отсортируем по убыванию.

In [231]:
print(data.groupby('family_status')['debt'].sum().sort_values(ascending=False)) 

family_status
женат / замужем          927
гражданский брак         385
Не женат / не замужем    273
в разводе                 84
вдовец / вдова            63
Name: debt, dtype: int64


На первый взгляд кажется, что зависимость между семейным положением и возвратом кредита в срок - имеется.

Категория Женатых/Замужних семей чаще допускают просрочки по кредитам, чем все остальные категории.

Проверим данное предположение рассмотрев конверсию данных.

In [232]:
print(data.groupby('family_status')['debt'].count().sort_values(ascending=False)) 
# для начала ознакомимся с общим количеством кредитов у каждой категории кредитополучателей.

family_status
женат / замужем          12261
гражданский брак          4134
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
Name: debt, dtype: int64


In [234]:
print(data.groupby('family_status')['debt'].mean())
# определяем конверсию данных для каждой категории кредитополучаетелей (для каждого семейного положения)

family_status
Не женат / не замужем    0.097639
в разводе                0.070648
вдовец / вдова           0.066246
гражданский брак         0.093130
женат / замужем          0.075606
Name: debt, dtype: float64


**Вывод:** 

1. Зависимость между семейным положением и возвратом кредита в срок - имеется!!

2. **Наибольшую долю просрочек** по кредитам в процентном соотношении совершает категория **Не женатых/Не замужних** лиц. (9,7%)

3. В то же время **Наименьшая доля просрочек** по кредитам у категории, относящейся к **Вдовам/Вдовцам**. (6,6%)


4. Несмотря на то, что наибольшее количество просрочек было у категории **Женатых/Замужних** кредитополучателей (927) - в процентном соотношении всего **7,6%** семей из данной категории совершают просрочки по кредитам.




Если считать, что любая разница между группами есть повод для «разделения» классов ...) кредитному отделу следует с осторожностью относиться к клиентам, которые не состоят или не состояли в браке. Согласно данным такие клиенты менее ответственные. Получается, что узаконенные отношения (включая и бывшие отношения) ведут к более ответственному отношению к своим обязательствам. Ну вроде как совпадает со здравым смыслом - семейные люди более скрупулёзно берегут свой семейное спокойствие и несут за него ответственность.



#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Для определния зависимости между уровнем дохода и возвратом кредита в срок воспользуемся ранее созданной категоризацией кредитополучателей.
Ранее мы разделили всех клиентов на 5 групп, отличающихся уровнем дохода.

Группы были поделены следующим образом:

Кредитополучатели чей доход составляет: 0–30000 — 'E';\
Кредитополучатели чей доход составляет: 30001–50000 — 'D';\
Кредитополучатели чей доход составляет: 50001–200000 — 'C';\
Кредитополучатели чей доход составляет: 200001–1000000 — 'B';\
Кредитополучатели чей доход составляет: 1000001 и выше — 'A'.

In [235]:
print(data.groupby('total_income_category')['debt'].sum().sort_values(ascending=False)) 
# группируем кредитополучателей по столбцу 'total_income_category'. Сортируем данные в порядке убывания.

total_income_category
C    1353
B     354
D      21
A       2
E       2
Name: debt, dtype: int64


Из полученных данных предполагаем, что в количественном выражении зависимость между уровнем дохода и возвратом кредита в срок присутствует.

Наибольшую долю просрочек осуществляют лица, входящие в категорию "С", т.е. те, чей доход составляет от 50001-200000.

При этом лица из категорий Е (чей доход до 30000) и А (чей доход более 1000001) - реже всего осуществляют просрочки по кредитам.

Для того, чтобы проверить данные утверждения, необходимо рассмотреть конверсию данных.

In [236]:
print(data.groupby('total_income_category')['debt'].count().sort_values(ascending=False)) 
# для начала ознакомимся с общим количеством кредитов у каждой категории кредитополучателей

total_income_category
C    15921
B     5014
D      349
A       25
E       22
Name: debt, dtype: int64


Пока что ничего удивительного. Чаще всего кредиты берут лица чей доход составляет от 50001-200000.

In [237]:
print(data.groupby('total_income_category')['debt'].mean()) 
# Определяем среднюю конверсию данных для каждой категории кредитополучателей

total_income_category
A    0.080000
B    0.070602
C    0.084982
D    0.060172
E    0.090909
Name: debt, dtype: float64


**Вывод:** 
В результате рассмотрения конверсии данных можно сделать однозначные выводы:

1. Зависимость между уровнем дохода и просрочкой по кредитам - имеется!

2. Не смотря на то, что в количественном выражении категория граждан "Е" с доходом (от 0-30000) реже всего берёт кредиты (всего 22 раза) - данная категория кредитополучаетей чаще допускает просрочки по кредитам (~9%).

3. Наибольшей группой кредитополучателей является категория граждан "C' с доходом (от 50001-200000).
8,4% граждан из данной категории допускают просрочки по кредитам, что является довольно-таки большим значением.

4. Наименьшее количество просрочек допускает категория "D" с доходом (от 30001–50000).

5. На мой взгляд удивительно, что категория граждан "А" с доходом (от 1 000 001) не только берет кредиты, но ещё умудряется допускать просрочки по данным кредитам кредитам. Доля просрочек у данной категории составляет целых 8%.

На мой взгляд - это самый сложный вопрос (для интерпретации результатов) в этом проекте. 

Быстрый взгляд на данные (в случае графика - это конечно ещё более наглядным становиться): ожидаемо, для группы с низким доходом (Е) процент невозвратов самый высокий. Что более интересно, так это то, что с ростом доходов тенденция (D->C и B->А) к снижению числа долгов не сохраняется, а скорее даже растёт. Можно сказать, что по какой-то причине люди "перестают считать деньги" или планировать "наперёд". Из-за подобного поведения наших групп параметр ("доход") нельзя назвать надёжным при определении платёжеспособности клиента. Без дополнительной информации он может даже наоборот привести к выдаче кредита состоятельному, но не слишком "надёжному" клиенту, хотя, возможно, это может быть аномалия, вызванная недостаточным размером выборки для клиентов с категорией доходов А.    
    
А теперь мой вывод- дополнительный, альтернативный вариант: исходя из здравого смысла логично предположить, что чаще всего должниками становятся люди с небольшим доходом, которым финансовые проблемы не позволяют вовремя осуществлять платежи. Но глядя на наши результаты лично мой вывод таков: выборка по заемщикам, как с наименьшим доходом, так и с наибольшим крайне недостаточна для формирования однозначных выводов. Можно думать, что E-категория самая низкооплачиваемая не может корректно распоряжаться деньгами и «влезает» в кредиты, но тогда неясно почему у А-категории также достаточно высокий процент задолженности. В общем вывод по группам А и Е неочевиден, слишком мало данных.

Если же взять для анализа только группы В, С и D, то по полученным данным не наблюдаем зависимости "те, кто получает больше ЗП, лучше возвращает кредит». В целом это логично, так как материальный достаток - не единственная переменная, которая влияет на качество кредитной истории.

p.s. Однако 2174 пропуска в данных столбца "total_income" ставят под сомнение полученный нами результат. Сводная таблица (или группировка) с данными о зависимости между уровнем дохода и возвратом кредита в срок возможно могла бы выглядеть иначе, если бы не пропущенные данные.


#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Для определения зависимости цели кредита на его возврат в срок воспользуемся ранее созданой категоризацией целей покупки, а именно столбцом 'purpose_category'.
Для начала просчитаем сумму просроченных кредитов по каждой категории "цели кредита".

In [238]:
print(data.groupby('purpose_category')['debt'].sum().sort_values(ascending=False)) 
# группируем по столбцу purpose_category. Сортируем по убыванию.

purpose_category
операции с недвижимостью    780
операции с автомобилем      400
получение образования       369
проведение свадьбы          183
Name: debt, dtype: int64


Из полученных данных предполагаем, что в количественном выражении зависимость между целью кредита и его возвратом в срок присутствует.

Кредиты взятые для целей операции с недвижимостью чаще допускают просрочку, чем кредиты взятые на проведение свадеб и кредиты для получения образования.

Но для более качественного анализа также как и раньше рассмотрим конверсию данных. (соотношение просроченных кредитов ко всем взятым).

In [239]:
print(data.groupby('purpose_category')['debt'].count().sort_values(ascending=False)) 
# ознакомимся с общим количеством взятых кредитов для каждой цели 

purpose_category
операции с недвижимостью    10751
операции с автомобилем       4279
получение образования        3988
проведение свадьбы           2313
Name: debt, dtype: int64


Опять же - ничего удивительного. Чаще всего берут кредиты для операций с недвижимостью.

In [240]:
print(data.groupby('purpose_category')['debt'].mean())
# Определяем среднюю конверсию данных для каждой категории цели кредита

purpose_category
операции с автомобилем      0.093480
операции с недвижимостью    0.072551
получение образования       0.092528
проведение свадьбы          0.079118
Name: debt, dtype: float64


**Вывод:** 

1. Цели кредита и вправду влиют на то, будут ли они возвращены в срок или нет.
2. **Наибольшая доля просрочек** приходится по кредитам, которые взяты для **операций с автомобилем** (Покупка, ремонт). **(9,3%)**
3. Вторым по опасности (по мнению Банка) будут кредиты для целей образования. По ним допускаются просрочки с вероятностью в **9,25%**
4. Несмотря на то, что наибольшее количество кредитов берутся для **операций с недвижимостью** - (10751), по данным целям кредита кредитополучатели **реже всего допускают просрочки**. Всего **(7,2%)** просрочек приходится на данный вид кредита.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

Для определения возможных причин появления пропусков в исходных данных попробуем ознакомится с самими данными.    
    Рассмотрев ещё раз проект, а именно: Раздел 2. Подраздел 2.1. задание №4. 
    Можем увидеть, что пропуски в данных были только в двух столбцах:
        
    -days_employed - общий трудовой стаж в днях
    -total_income - ежемесячный доход.
Начнем мини-исследование:\
Прежде, чем изучить данные, попробуем составить логические предположения, почему в данных столбцах могут отсутствовать данные?\
Одно из предположений может звучать так:\
**"отсутствие данных в столбцах связано с тем, что, вероятно, у данных кредитополучателей трудовой стаж отсутствует совсем".**\
т.е. другими словами - данный клиент:\
-либо никогда в жизни не работал.\
-либо же никогда не работал "Официально".\
Второе предположение звучит так:\
**"отсутсвие данных в столбах связано с тем, что, вероятно, сотрудники банка или клиенты забыли внести или предоставить необходимые данные"**\
в таком случае, вероятно в данных мы увидим, что данные по стажу будут отсутствовать в одних строках, а данные по ежемесячному доходу будут отсутствовать в других строках.

Для проверки данных гипотез попробуем вывести все строки датафрейма, содержащие отсутствующие значения (значеия - NaN).\
Данное мини-исследование проведем как раз в Разделе 2. Подразделе 2.1. После задания №4, до момента, когда эти отсутствующие данные мы заменим медианными значениями. (см. выше по проекту).

Написав дополнительную функцию и получив строки с отсутствующими значениями мы обнаружили интересные закономерности.

1. **Отсутствие данных в столбцах 'days_employed' и 'total_income' полностью взаимосвязано!**.

В случае, если бы эти столбцы не были взаимосвязаны - можно было бы предположить, что кредитополучатель просто напрасто забыл предоставить сотруднику банку одну из данных справок или предоставил не подтвержденную справку о доходах или стаже.

Однако в нашем случае данные столбцы взаимосвязаны.

Таким образом можно сделать логическое предположение, что графы 'days_employed' и 'total_income' - заполняются сотрудниками банка только в том случае, если кредитополучатель официально подтвердил свой трудовой стаж и наличие доходов.

1) Если бы данные кредитополучатели были бы *'безработными'*(и соответственно без доходов) - в таком случае у них в столбце income_type стояла бы пометка 'безработный'.\
Однако, как мы видим данное предположение *не подтверждается*.\
Если рассмотреть данные подробнее, можно увидеть, что столбец 'income_type' у данных клиентов *не соответствует* статусу 'безработный'.

2) Второе предположение, можно сделать такое:\
Отсутствие данных в столбце 'total_income' связано с отсутствием данных в столбце 'days_employed', потому что, данные кредитополучатели *никогда в жизни не работали*.\
Чтобы проверить данную теорию вновь ознакомимся с той же таблицей.\
Как мы можем увидеть, НЕТ определенно четкой закономерности по возрасту у тех клиентов, у которых отсутствуют данные в этих столбцах.(т.е. есть и молодые люди и люди возрастной категории)\
соответственно сделать предположение, что данные кредитополучатели никогда в жизни не работали - *нельзя*!

3) Последнее предположение звучало бы так:\
Вероятно отсутствие данных связано с тем, что данные клиенты ***никогда не работали 'Официально'.***\
В таком случае понятно, почему столбец с описанием общего трудового стажа отсутствует у данных клиентов, и в прицнипе, почему графа ежемесячного дохода для этих клиентов также пуста.\
Вероятно данные клиенты получают зарплату - 'Не официально' и, следовательно, хоть и имеют ежемесячый доход, они данный доход не могут подтвердить официально.(т.е. условно говоря, они получают зарплату "в конвертах").

Объединяя всё выше описанное можно попробовать сделать окончательное предположение:

**Проведя исследование, мы выяснили что отсутствие значений доходов и стажа у некоторых клиентов:**\
***1. Нельзя отнести к тому, что все данные клиенты являются безработными*** (иначе в графе стоял бы статус безработного).\
***2. Вероятно, отсутствие этих данных связано с тем, что данные кредитополучатели работают 'Не официально'.***\
(т.е. не могут официально подтвердить свой трудовой стаж и свой ежемесячный доход)

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Среднее значение некорректно характеризует данные, когда некоторые значения сильно выделяются среди большинства. 

Можно взять пример из теории:
Например, 5 сотрудников компании получают зарплату 100 000 рублей. 

Инвестор пригласил директора по маркетингу с зарплатой 300 000 рублей.

Если считать среднюю зарплату - она составит 133 333,33 рублей, хотя по факту, зарплата у 5 сотрудников осталась той же. (100000).

При расчете медианы - таких проблем не возникает, так как медиана рассчитает среднее из "срединных значений". 

[100, 100, 100, 'MEDIANA' 100, 100, 300]

Вывод: для количественных переменных медианное значение подойдет лучше, так как не будет учитывать сильно выделяющиеся (странные) значения.

### Шаг 4: общий вывод.

**Вывод**

В данном проекте производилось исследование надежности заемщиков Банка.

Нами была рассмотрена таблица с данными о кредитополучателях.

Произведя нехитрые операции по обработке данных - нам удалось осуществить некоторые наблюдения, и на их основании составить некоторые утверждения.
Данные утверждения в теории в будущем могут помочь составить кредитный рейтинг потенциального кредитополучателя.

Обращается особое внимание, что на эти данные, как и на любые другие данные (в век цифровых технологий), можно надеяться только в определенной степени, так как они не могут учитывать всех обстоятельств каждого по отдельности кредитополучателя, а также не могут учитывать надежность будущих потенциальных кредитополучателей.

Также обращается внимание, что полученные данные и сделанные на их основе выводы актуальны только в момент проведения данных измерений. В дальнейшем выводы могут быть другими.

Учитывая всё вышеописанное можно привести некоторые из сделанных выводов, полученных в ходе исследования:

1. Возвраты кредитов в срок зависят от многих обстоятельств, в том числе:
\
    1.1. От количества детей у кредитополучателей.\
    1.2. От семейного положения кредитополучателей.\
    1.3. От уровня дохода у кредитополучателей.\
    1.4. От целей получения кредита и т.д.


2. Было определено, что в **процентном** соотношении - кредитополучатели имеющие детей, с большой вероятностью будут допускать просрочки по платежам, нежели кредитополучатели, не имеющие детей.
При этом в **количественном** выражении, именно кредитополучатели без детей совершили больше всего просрочек по платежам.


3. Те же пропорции сохраняются и для критерия семейного положения кредитополучателей.
Наибольшее **количество** пропущенных платежей по кредитам приходится на категорию Женатых/Замужних граждан.
При этом в **процентном** соотношении самыми неблагополучными в плане возврата кредитов в срок - являются Не Женатые/Не замужние граждане. **(9,7%).**


4. Не смотря на то, что в **количественном** выражении категория граждан с доходом (от 0-30000) реже всего берёт кредиты (всего 22 раза) - данная категория кредитополучаетей чаще допускает просрочки по кредитам **(~9%)**.
Наибольшей группой кредитополучателей является категория граждан с доходом (от 50001-200000). 8,4% граждан из данной категории допускают просрочки по кредитам, что является довольно-таки большим значением.
Наименьшее количество просрочек допускает категория граждан с доходом (от 30001–50000).


5. Также, было определено, что **наибольшая доля просрочек** приходится по кредитам, которые были взяты для **операций с автомобилем** (Покупка, ремонт). **(9,3%)**
Вторым по неблагонадежности для Банка будут кредиты для целей образования. По ним допускаются просрочки с вероятностью в 9,25%
Несмотря на то, что наибольшее количество кредитов берутся для **операций с недвижимостью** - (10751), по данным целям кредита кредитополучатели **реже всего допускают просрочки**. Всего **(7,2%)** просрочек приходится на данный вид кредита.


Для составляения правильного кредитного потенциала для заемщиков этих данных недостаточно.
Однако, как можно было заметить, из полученных данных можно выявить некоторые закономерности, которые потенциально можно будет использовать при следующих исследованиях.


***Данное исследование подтвердило наше предположение, что с данными можно и нужно работать и в них можно отслеживать интересные закономерности, которые потенциально могут принести прибыль компании или же спасти её от потенциальных убытков.***

С уважением,


Степанян Ваган Сергеевич